components/raftstore/src/store/async_io/write.rs

```diff
@@ -0,0 +1,649 @@
+                .consume_and_shrink(
+                    &mut self.batch.raft_wb,
+                    true,
+                    RAFT_WB_SHRINK_SIZE,
+                    RAFT_WB_DEFAULT_SIZE,
+                )
+                .unwrap_or_else(|e| {
+                    panic!(
+                        "store {}: {} failed to write to raft engine: {:?}",
+                        self.store_id, self.tag, e
+                    );
+                });
+            self.perf_context.report_metrics();
+            STORE_WRITE_RAFTDB_DURATION_HISTOGRAM
+                .observe(duration_to_sec(now.saturating_elapsed()) as f64);
+        }
+
+        self.batch.after_write_to_raft_db(&self.metrics);
+
+        fail_point!("raft_before_follower_send");
+
+        let now = Instant::now();
+        let mut unreachable_peers = HashSet::default();
+        for task in &mut self.batch.tasks {
+            for msg in task.messages.drain(..) {
+                let msg_type = msg.get_message().get_msg_type();
+                let to_peer_id = msg.get_to_peer().get_id();
+                let to_store_id = msg.get_to_peer().get_store_id();
+                if let Err(e) = self.trans.send(msg) {
+                    // We use metrics to observe failure on production.
+                    debug!(
+                        "failed to send msg to other peer in async-writer";
+                        "region_id" => task.region_id,
+                        "peer_id" => task.peer_id,
+                        "target_peer_id" => to_peer_id,
+                        "target_store_id" => to_store_id,
+                        "err" => ?e,
+                        "error_code" => %e.error_code(),
+                    );
+                    self.message_metrics.add(msg_type, false);
+                    // Send unreachable to this peer.
+                    // If this msg is snapshot, it is unnecessary to send snapshot
+                    // status to this peer because it has already become follower.
+                    // (otherwise the snapshot msg should be sent in store thread other than here)
+                    unreachable_peers.insert((task.region_id, to_peer_id));
+                } else {
+                    self.message_metrics.add(msg_type, true);
+                }
+            }
+        }
+        self.trans.flush();
+        self.message_metrics.flush();
+        for (region_id, to_peer_id) in unreachable_peers {
+            self.notifier.notify_unreachable(region_id, to_peer_id);
+        }
+        let now2 = Instant::now();
+        STORE_WRITE_SEND_DURATION_HISTOGRAM
+            .observe(duration_to_sec(now2.saturating_duration_since(now)));
+
+        for (region_id, (peer_id, ready_number)) in &self.batch.readies {
+            self.notifier
+                .notify_persisted(*region_id, *peer_id, *ready_number, now);
+        }
+        STORE_WRITE_CALLBACK_DURATION_HISTOGRAM.observe(duration_to_sec(now2.saturating_elapsed()));
```

Comment:
It seems to only send several messages, is it necessary?

Reply:
Sending persisted msgs is not very fast. In sysbench insert workload tests, the time it takes is about half of handling write tasks, 1.5 times of sending raft msgs.